## 01. Import libraries

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Turn Project Folder into Path
path = r'/Users/tanayawinder/Desktop/6-27-2024 Instant Cart Basket Analysis'

In [4]:
# Check output
path

'/Users/tanayawinder/Desktop/6-27-2024 Instant Cart Basket Analysis'

In [5]:
# Import orders.csv (Original Data) via OS library
orders = pd.read_csv(os.path.join(path,'02 Data','Prepared Data', 'orders_wrangled.csv'),index_col = False)

In [6]:
# Import products.csv (Original Data) via OS library
departments = pd.read_csv(os.path.join(path,'02 Data','Prepared Data', 'deparments_wrangled.csv'))

In [7]:
# Import products.csv (Original Data) via OS library
products = pd.read_csv(os.path.join(path,'02 Data','Original Data', 'products.csv'))

## 02. Data Consistency Checks

### Finding Missing Values

In [10]:
products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Examine missing values
df_nan = products[products['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
products.shape

(49693, 5)

In [14]:
prods_clean = products[products['product_name'].isnull() == False]

In [15]:
prods_clean.shape

(49677, 5)

### Address Duplicates

In [17]:
df_dups = prods_clean[prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
prods_clean.shape

(49677, 5)

In [20]:
prods_clean_no_dups = prods_clean.drop_duplicates()

In [21]:
prods_clean_no_dups.shape

(49672, 5)

In [22]:
# Export checked products df
prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 4.5 Task

In [24]:
orders.describe()

Unnamed: 0      order_id       user_id  number_of_orders  \
count  3.421083e+06  3.421083e+06  3.421083e+06      3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05      1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04      1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00      1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04      5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05      1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05      2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05      1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [25]:
# Set the display option to suppress scientific notation
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [26]:
orders.describe()

Unnamed: 0   order_id    user_id  number_of_orders  order_day_of_week  \
count  3421083.00 3421083.00 3421083.00        3421083.00         3421083.00   
mean   1710541.00 1710542.00  102978.21             17.15               2.78   
std     987581.74  987581.74   59533.72             17.73               2.05   
min          0.00       1.00       1.00              1.00               0.00   
25%     855270.50  855271.50   51394.00              5.00               1.00   
50%    1710541.00 1710542.00  102689.00             11.00               3.00   
75%    2565811.50 2565812.50  154385.00             23.00               5.00   
max    3421082.00 3421083.00  206209.00            100.00               6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

There are less entries (count) in days_since_prior_order than in the other columns. As there is a 0 recorded as the minimum, it is likely that there are blanks (missing) instead of blanks meaning "zero days". These missing values should be investigated. In user_id column, the max is less than the count. As this is not a numerical column, this discrepency should be investigated further. 

In [28]:
orders.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
number_of_orders            int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [29]:
orders['user_id'] = orders['user_id'].astype(str)

In [30]:
orders['order_id'] = orders['order_id'].astype(str)

In [31]:
orders.dtypes

Unnamed: 0                  int64
order_id                   object
user_id                    object
number_of_orders            int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [32]:
orders.describe()

Unnamed: 0  number_of_orders  order_day_of_week  order_hour_of_day  \
count  3421083.00        3421083.00         3421083.00         3421083.00   
mean   1710541.00             17.15               2.78              13.45   
std     987581.74             17.73               2.05               4.23   
min          0.00              1.00               0.00               0.00   
25%     855270.50              5.00               1.00              10.00   
50%    1710541.00             11.00               3.00              13.00   
75%    2565811.50             23.00               5.00              16.00   
max    3421082.00            100.00               6.00              23.00   

       days_since_prior_order  
count              3214874.00  
mean                    11.11  
std                      9.21  
min                      0.00  
25%                      4.00  
50%                      7.00  
75%                     15.00  
max                     30.00

In [33]:
# Check for mixed data types
for col in orders.columns.tolist():
  weird = (orders[[col]].map(type) != orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (orders[weird]) > 0:
    print (col)

There are no mixed data types in orders.csv

In [35]:
orders.shape

(3421083, 7)

In [36]:
# Check orders CSV for nulls

orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
number_of_orders               0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 nulls, all in the days_since_prior_order column.`

In [38]:
#Investigating blanks

orders['days_since_prior_order']

0           NaN
1         15.00
2         21.00
3         29.00
4         28.00
           ... 
3421078   29.00
3421079   30.00
3421080   18.00
3421081    7.00
3421082   30.00
Name: days_since_prior_order, Length: 3421083, dtype: float64

In [39]:
#Investigating blanks - null table

df_nan_ords =  orders[orders['days_since_prior_order'].isnull()==True]

In [40]:
df_nan_ords

Unnamed: 0 order_id user_id  number_of_orders  order_day_of_week  \
0                 0  2539329       1                 1                  2   
11               11  2168274       2                 1                  2   
26               26  1374495       3                 1                  1   
39               39  3343014       4                 1                  6   
45               45  2717275       5                 1                  3   
...             ...      ...     ...               ...                ...   
3420930     3420930   969311  206205                 1                  4   
3420934     3420934  3189322  206206                 1                  3   
3421002     3421002  2166133  206207                 1                  6   
3421019     3421019  2227043  206208                 1                  1   
3421069     3421069  3154581  206209                 1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

All null entries in days_since_prior_order have '1' listed under number_of_orders. These entries are likely for a customers first purchase, and thus there is no available date of prior order.


In this case, a null entry is as informative as a numerical entry. This column should remain in the dataset and no numbers should be input.


In [43]:
df_dups_ords = orders[orders.duplicated()]

In [44]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, number_of_orders, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates in orders.csv.

## Export

In [47]:
orders.shape

(3421083, 7)

In [48]:
orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))